rids/rids.py

```diff
@@ -14,114 +14,139 @@
-        for line in ioc_data.readlines():
-          line = line.strip().decode('utf-8')
-          if not line: continue
-          bad_ip = str(ipaddress.ip_address(line))
-          if bad_ip not in bad_ips:
-            bad_ips[bad_ip] = [source['name']]
-          else:
-            bad_ips[bad_ip].append(source['name'])
-  return bad_ips
-
-
-def get_external_ip():
-  with urllib.request.urlopen('https://ipinfo.io/ip') as my_ip:
-    return my_ip.read().decode("utf-8").strip()
+                    'The IP of the host process, helps determine whether an IP '
+                    'is a client or a remote endpoint.')
+flags.DEFINE_string('config_path', '/etc/rids/config.json',
+                    'file path indicating where IOC configuration can be found')
+flags.DEFINE_string('eventlog_path', None,
+                    'file path indicating where to append new event logs')
 
 
 def main(argv):
+  """Main entry point of the app.  Also bound to CLI `rids` by setuptools."""
   if len(argv) > 1:
     raise app.UsageError('Too many command-line arguments.')
 
+  eventlog_path = _get_eventlog_path()
+  logging.basicConfig(level=logging.DEBUG,
+                      format='%(asctime)s %(levelname)-8s %(message)s',
+                      datefmt='%m-%d %H:%M',
+                      filename=eventlog_path,
+                      filemode='a')
+
+  config = _load_config()
+  ruleset = iocs.fetch_iocs(config)
+
+  host_ip = _get_host_ip()
+  print(f'Using {host_ip} as host IP address')
+
+  event_queue = queue.Queue()
+  loop = asyncio.get_event_loop()
+  with ProcessPoolExecutor(max_workers=3) as executor:
+    loop.run_in_executor(
+        executor,
+        partial(_inspect_tls_traffic, host_ip, ruleset, event_queue))
+    loop.run_in_executor(
+        executor,
+        partial(_inspect_remote_endpoints, host_ip, ruleset, event_queue))
+  
+    while True:
+      # We just log the suspicious events for now, but here is where we could
+      # do post-processing and/or share sightings of known / unknown threats.
+      event = event_queue.get()
+      logging.log(json.puts(event))
+      event_queue.task_done()
+
+
+def _inspect_tls_traffic(host_ip, ruleset, q):
+  """Worker function for thread that inspects client and server hellos.
+  
+  Args:
+    ruleset: a RuleSet object to perform evaluation with
+    q: a queue.Queue for relaying events back to the main thread
+  """
+  tls_connection = TlsConnectionMonitor(host_ip)
+  for tls_info in tls_connection.monitor():
+    events = ruleset.match_tls(tls_info)
+    for event in events:
+      q.put(event)
+
+
+def _inspect_remote_endpoints(host_ip, ruleset, q):
+  """Worker function for thread that inspects remote IP addresses.
+
+  Args:
+    ruleset: a RuleSet object to perform evaluation with
+    q: a queue.Queue for relaying events back to the main thread
+  """
+  remote_ip = IpPacketMonitor(host_ip)
+  for ip_info in remote_ip.monitor():
+    events = ruleset.match_ip(ip_info)
```

Comment:
This will probably read like `ruleset.ip_matcher.match_ip(ip_info)`

Reply:
done